### ISSUE TITLE
New Zone Charts - BABIP, wOBA, etc.

### ISSUE THREAD

**Issue by kruser** [kruser/pitchfx-site#54]
I'd like to add some new heatmaps to the zone charts we already have. Here is what exists...
![image](https://cloud.githubusercontent.com/assets/2208700/2587147/63dc01a4-ba16-11e3-9c28-14897662ade2.png)

For this issue, I'd like to add..
- wOBA
- BABIP
- BIP

I'll leave this unassigned for now. If anybody is interested in adding these I can walk you through the heatmaps framework I setup for the others.

**Comment by albertlyu** [comment]
If it's alright, I'd like to give this a shot. Will let me get back on this project right where I left off, and also may give me a better idea of how to approach #39 in terms of what's possible now that wasn't before. Feel free to assign this issue to me and I will take a look at it this week and during the weekend. Pointing me in the right direction in terms of which lines of code in which js files to add new objects should be a good start. Thanks!

**Comment by kruser** [comment]
@albertlyu, yes, go ahead and start it. I'm unable to assign the issues to people that don't have push access to the repository at the moment, so I'll just stick an 'in-progress' label on the issue.

Tech details...

The UI picks up the definitions of the charts here.
https://github.com/kruser/pitchfx-site/blob/master/app/app/scripts/controllers/pitchStats.js#L19-L84

I would be fine with these new charts being at the top of the list.

You'll see that each chart definition has a "generator" function. If you unwind that, it simply calls into Zones.js, which calls into it's 2D array of Zone.js objects.

The function you add in Zone.js will do the math on all pitches that have been assigned to that particular zone. See this swingRate example:
https://github.com/kruser/pitchfx-site/blob/master/app/app/scripts/pojos/Zone.js#L80-L105

Tests would also be welcome. From the "app" folder you simply need to run

```
karma start
```

Doing so keeps the Karma test runner going, watching your changes. Here is the tests for the swing rate function.
https://github.com/kruser/pitchfx-site/blob/master/app/test/spec/pojos/Zones.js#L110-L184

Note that the Zone.js functions return a ZoneStat.js object. The value and description are used in the chart and tooltip. Also note that the Zone.js has an array of Pitch.js objects, which contains helper methods for balls in play, some wOBA stuff, etc., etc.
https://github.com/kruser/pitchfx-site/blob/master/app/app/scripts/pojos/Pitch.js

Any new methods to Pitch.js probably need tests as well.
https://github.com/kruser/pitchfx-site/blob/master/app/test/spec/pojos/Pitch.js

**Comment by albertlyu** [comment]
Great explanations. Any rhyme or reason in choosing the max for each heatmap? I'm finding Kershaw's BIP rate doesn't split nicely between max: 0.400 and max: 0.500 for bip / pitches. If it's arbitrary, then that's totally fine and I'll probably go with 0.500 :)

![image](https://cloud.githubusercontent.com/assets/1390356/2610711/45e1cc50-bb8e-11e3-8f3e-3208941a8ed5.png)

**Comment by kruser** [comment]
The max for each chart should be whatever we consider high for that stat, I.e. whatever we'd want to see red. I eyeballed the others. I would think we should aim for the 10th percentile or something like that.

**Comment by albertlyu** [comment]
So double-checking here, but pitch.isHit() includes home runs, yes? Based on https://github.com/kruser/pitchfx-site/blob/master/app/app/scripts/pojos/Pitch.js#L56-L62 I'm asking because BABIP is typically hits _excluding home runs_ divided by balls in play.

**Comment by kruser** [comment]
Yes, it includes home runs. You should add some functions to get singles, doubles, etc. See getWeightedObavalues for examples, although you won't need that function directly.

Also, we don't have the proper coefficients in the UI code yet so you can punt on that for now and just use 2014 values.

**Comment by kruser** [comment]
BTW, these are things that the data model should have (singles, doubles etc). Looking at these improvements to atbat-mongodb so we'll have less regex parsing here.

**Comment by albertlyu** [comment]
Right, makes sense. Because currently these des events are at the atbat-level, not at the pitch-level, correct? Are those the changes you're planning to improve on atbat-mongodb? Adding these atbat-level attributes intelligently to individual pitches will render many of the functions in Pitch.js redundant, as I understand it.

**Comment by kruser** [comment]
Exactly. Too much for a person writing UI to need to know.

**Comment by albertlyu** [comment]
Something like this? Seems to work for me, though I'd need to find an example of a home run that was omitted from the numerator. I know the other direction would be to add functions to return boolean values if single, if double, etc. but I wanted to see if I understand correctly how the UI code uses the data model.

```
/**
 * @returns {Boolean} true if this pitch resulted in a hit, excluding home runs
 */
pitchfx.Pitch.prototype.isHitNoHR = function()
{
    if (this.hip && this.hip.des)
    {
        var des = this.hip.des;
        if (/home run/i.test(des))
        {
            return false;
        }
        else
        {
            return (/in play,\s*(no out|run.*)/i.test(this.des));
        }
    }
    return undefined;
};
```

**Comment by kruser** [comment]
Don't repeat yourself though. You can call other functions on the prototype like this.isHit().

I tend to think you should just add a isHomeRun() function and && the two together from the caller.

(isHit() && !isHomeRun())

**Comment by albertlyu** [comment]
Sounds good. The DRY principle is one you can feel free to repeatedly drill into my head.

Thinking about wOBA on my commute home. wOBAvalue / bip makes the most sense to me for each zone, not wOBAvalue / pitch. Let me know if you disagree.

**Comment by kruser** [comment]
What about a strikeout? Should we treat a whiff as a strikeout?

**Comment by albertlyu** [comment]
I hesitate to do that. For a strikeout and a walk, which pitch do we attribute the wOBA value to? The 3rd strike or the 4th ball? This opens up a can of worms discussion (or a fascinating discussion, depending on your point of view) leading to pitch values, the assumption that pitches are iid, and pitch sequencing. 

At least for balls in play, we can confidentially assign the wOBA value to the pitch that resulted in the ball in play. The way you set up the getweightedOBAvalue() led me to believe that the end result is that we only assign wOBA coefficients to balls in play that resulted in the batter on base.

**Comment by kruser** [comment]
From my point of view, you can't calculate atbat based things like wOBA or Batting Average based on single pitches. So we're thinking the same thing.

I'm not sure what is on brooksbaseball.net profile charts - http://www.brooksbaseball.net/h_profile.php?player=408045&gFilt=&pFilt=FA|SI|FC|CU|SL|CS|KN|CH|FS|SB&time=month&minmax=ci&var=baa&s_type=13&startDate=03/30/2007&endDate=04/03/2014&balls=-1&strikes=-1&b_hand=-1

So I'm cool with wOBA/BIP. Just make it so that the title of the chart is descriptive and not simply wOBA. I think I'll also add a subtitle to the charts too after your work so we can squeeze more info in.

As for the getWeightedObaValue(), I used that when I was doing the bubble chart, as the wOBA value was the radius of the bubble too. I removed that chart in favor of these heatmaps though. I'm not sure you can use it as-is.

Looking good though. I'm excited for these new charts!

**Comment by albertlyu** [comment]
For the brooksbaseball.net charts, the denominator is not pitches, but neither does it appear to be BIP (by comparing with the BABIP views and HR / BIP views). Anyway, for wOBA, I'll go with wOBA / BIP and I was able to use getWeightedObaValue() as is, but I'll let you know if I run into problems there while testing.

For BABIP, I like the (isHit() && !isHomeRun()) approach, as I can similarly do (isBallInPlay() && !isHomeRun()) for the denominator. I'll give these a shot.

**Comment by albertlyu** [comment]
Hmm, so I'm not exactly sure what's an efficient way to test the new functions getWOBA(), getBABIP(), and getBIPRate() in Zones.js and isHomeRun() in Pitch.js. How do you grab JSON for a test pitch that Josmil Pinto has seen that matches the criteria? Do you manually query the Mongo database to get such an at-bat?

Anyway, I figured I can submit a pull request first for you to review the changes I made, and/or I can run karma tests once I figure out how to grab an example pitch in JSON format easily.

**Comment by kruser** [comment]
On chrome/webkit you can inspect-element and look at the network trace to the pitches API. You'll see calls like this one go through:
http://baseballmod.com/api/pitches/500887?atbatFilter=%7B%22playerCard%22:%22batter%22,%22pitcherHand%22:%22%22,%22batterHand%22:%22%22,%22date%22:%7B%22start%22:%222013-01-01%22,%22end%22:%222014-04-03%22%7D,%22runners%22:%7B%22gate%22:%22OR%22%7D,%22outs%22:%7B%7D,%22balls%22:%7B%7D,%22strikes%22:%7B%7D,%22gameType%22:%7B%22R%22:true%7D%7D

I snip pitches from there and plug them into tests, alter the px and pz coordinates so they drop in the expected zones.

Also, if you use Chrome and are testing the json/RESTful APIs I recommend getting this plugin for Chrome.
https://chrome.google.com/webstore/detail/json-formatter/bcjindcccaagfpapjjmafapmmgkkhgoa?hl=en

**Comment by albertlyu** [comment]
I have to say, these karma tests you set up are very slick. In the original isHomeRun() definition in commit 6217fc0813a88a539e5980b8258b2904f35ce07d, I was looking for the property "home run" in the wrong des object. I really should have been looking in this.hip.des instead. By running karma tests and testing on the non-HR plays, I realized I needed to check if this.hip.des was defined, and if not, then return undefined. Not only that, but I have to expect undefined for the non-hip pitches, and expect false for the non-HR hip pitches.

So instead of:

```
pitchfx.Pitch.prototype.isHomeRun = function()
{
    return (/home run/i.test(this.des));
};
```

I'm going with:

```
pitchfx.Pitch.prototype.isHomeRun = function()
{
    if (this.hip && this.hip.des)
    {
        return (/home run/i.test(this.hip.des));
    }
    return undefined;
};
```

Good stuff. Something to note for newer developers.

**Comment by kruser** [comment]
One more nit pick - functions starting with "is" should return a simple boolean always. Can you change the "return undefined" to "return false"? I realize both are "falsy" but "return false" is clearer.

**Comment by albertlyu** [comment]
Sounds good. I'll update the isHomeRun() function so that it returns false for both falsy scenarios, and adjust the expected boolean values in test Pitch.js.

**Comment by kruser** [comment]
The regex test is good. It already returns true or false. I was just talking about the fall through return at the end.

Let me know when the pull request is complete. I'll be able to update the site tomorrow afternoon if its ready.

Thanks again. Great work.

**Comment by albertlyu** [comment]
I think we're on the same page, but just to be clear:

```
pitchfx.Pitch.prototype.isHomeRun = function()
{
    if (this.hip && this.hip.des)
    {
        return (/home run/i.test(this.hip.des)); // returns true if hip is HR, false if hip is not HR
    }
    return false; // returns false if pitch is not hip
};
```

**Comment by kruser** [comment]
Perfect

**Comment by kruser** [comment]
@albertlyu, I merged your pull request in and made two slight modifications. Let me know if you have questions.
https://github.com/kruser/pitchfx-site/commit/4e64b155e186654880114cec85797a53d031d4ca
- changed the max wOBA to .500. Since we don't have all wOBA values in our calculation we're trending a bit high on wOBA. Guys like Aaron Hicks show up with too many hot zones. Setting the max up to .500 brings it back closer to reality. I think we're going to have to get these max values a little more scientifically soon.
- I changed the tooltip a little bit, removing the cumulative wOBA values since I think that just confuses.

Let me know if you're cool with those and I'll update the live site.
